keccak256/src/permutation/iota.rs

```diff
@@ -0,0 +1,200 @@
+use crate::arith_helpers::{convert_b2_to_b13, convert_b2_to_b9, A4};
+use crate::common::{PERMUTATION, ROUND_CONSTANTS};
+use crate::gate_helpers::biguint_to_f;
+use eth_types::Field;
+use halo2_proofs::circuit::AssignedCell;
+use halo2_proofs::circuit::Layouter;
+use halo2_proofs::{
+    plonk::{Advice, Column, ConstraintSystem, Error, Fixed, Selector},
+    poly::Rotation,
+};
+use itertools::Itertools;
+use num_bigint::BigUint;
+use std::convert::TryInto;
+
+#[derive(Clone, Debug)]
+pub struct IotaConfig<F> {
+    q_enable: Selector,
+    lane00: Column<Advice>,
+    flag: Column<Advice>,
+    round_constant: Column<Fixed>,
+    round_constant_b13: F,
+    round_constants_b9: [BigUint; PERMUTATION],
+}
+
+impl<F: Field> IotaConfig<F> {
+    /// Iota step adds a round constant to the first lane.
+    ///
+    /// We enable the gate to handle 3 different cases:
+    ///
+    /// The first case takes place in the first 23 rounds, the prover MUST add
+    /// the `A4` times round constant in base 9. We enforce it by requiring the
+    /// flag equal to one.
+    ///
+    /// The second the third cases happen in the 24-th
+    /// round. It depends if prover wants to absorb new input or not, which is
+    /// indicated by the flag.
+    ///
+    /// If prover doesn't want to absorb new input,
+    /// then add `A4 * round_constant_b9` as the previous 23-th round did.
+    ///
+    /// Otherwise, apply the round constant in base 13 to the state, which has
+    /// been mixed with new input and converted form base 9 to base 13.
+    pub fn configure(
+        meta: &mut ConstraintSystem<F>,
+        lane00: Column<Advice>,
+        flag: Column<Advice>,
+        round_constant: Column<Fixed>,
+    ) -> Self {
+        let q_enable = meta.selector();
+        meta.enable_equality(lane00);
+        meta.enable_equality(flag);
+
+        meta.create_gate("iota", |meta| {
+            let q_enable = meta.query_selector(q_enable);
+            let flag = meta.query_advice(flag, Rotation::cur());
+            let lane00_next = meta.query_advice(lane00, Rotation::next());
+            let lane00 = meta.query_advice(lane00, Rotation::cur());
+            let round_constant = meta.query_fixed(round_constant, Rotation::cur());
+            vec![q_enable * (lane00_next - lane00 - flag * round_constant)]
+        });
+
+        let round_constant_b13 =
+            biguint_to_f::<F>(&convert_b2_to_b13(ROUND_CONSTANTS[PERMUTATION - 1]));
+
+        let round_constants_b9: [BigUint; 24] = ROUND_CONSTANTS
+            .iter()
+            .map(|&x| convert_b2_to_b9(x))
+            .collect_vec()
+            .try_into()
+            .unwrap();
+
+        Self {
+            q_enable,
+            lane00,
+            flag,
+            round_constant,
+            round_constant_b13,
+            round_constants_b9,
+        }
+    }
+
+    /// The first 23 rounds
```

Comment:
Can you extend a bit the docs here as done below?
```suggestion
    /// The first 23 rounds. (No mixing logic involved).
    ///
    /// Applies IotaB9 steady-step logic.
    /// It consists of: `new_lane_00 - (lane00 * ROUND_CTANTS[round]) == 0`.
```